File: backend/app/schemas/incident.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class UserBrief(BaseModel):
    id: UUID
    full_name: str
    email: str
    avatar_initials: str | None

    model_config = {"from_attributes": True}


class ExternalRefOut(BaseModel):
    id: UUID
    external_source: str
    external_ref: str
    external_url: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ExternalRefCreate(BaseModel):
    external_source: str
    external_ref: str
    external_url: str | None = None


class IncidentCreate(BaseModel):
    title: str
    severity: Literal["sev1", "sev2", "sev3", "sev4"] = "sev2"
    template_id: str | None = None
    assigned_to: str | None = None


class IncidentUpdate(BaseModel):
    title: str | None = None
    severity: Literal["sev1", "sev2", "sev3", "sev4"] | None = None
    status: Literal["open", "contained", "closed", "monitoring"] | None = None
    executive_summary: str | None = None
    attack_vector: list[str] | None = None
    affected_users: int | None = None
    assigned_to: str | None = None
    metadata: dict | None = None


class IncidentOut(BaseModel):
    id: UUID
    org_id: UUID
    incident_ref: str
    title: str
    severity: str
    status: str
    template_id: UUID | None
    assigned_to: UUID | None
    assigned_user: UserBrief | None = None
    created_by: UUID | None
    opened_at: datetime
    contained_at: datetime | None
    closed_at: datetime | None
    executive_summary: str | None
    attack_vector: list[str]
    affected_users: int
    metadata: dict
    created_at: datetime
    updated_at: datetime
    external_refs: list[ExternalRefOut] = []

    model_config = {"from_attributes": True}

    @classmethod
    def model_validate(cls, obj, *args, **kwargs):
        # Map metadata_ ORM column to metadata field
        if hasattr(obj, "metadata_"):
            obj.__dict__["metadata"] = obj.metadata_
        return super().model_validate(obj, *args, **kwargs)


class IncidentStats(BaseModel):
    timeline_count: int
    ioc_count: int
    task_total: int
    task_done: int
    attachment_count: int
    duration_hours: float | None

File: frontend/src/types/incident.ts
export type Severity = 'sev1' | 'sev2' | 'sev3' | 'sev4'
export type IncidentStatus = 'open' | 'contained' | 'closed' | 'monitoring'

export interface UserBrief {
  id: string
  full_name: string
  email: string
  avatar_initials: string | null
}

export interface IncidentExternalRef {
  id: string
  incident_id: string
  external_source: string
  external_ref: string
  external_url: string | null
}

export interface Incident {
  id: string
  org_id: string
  incident_ref: string
  title: string
  severity: Severity
  status: IncidentStatus
  assigned_to: string | null
  assigned_user: UserBrief | null
  executive_summary: string | null
  notes: string | null
  lessons_learned: string | null
  actions_todo: string | null
  attack_vector: string[]
  affected_users: number
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
  contained_at: string | null
  closed_at: string | null
  external_refs: IncidentExternalRef[]
}

export interface IncidentStats {
  timeline_count: number
  ioc_count: number
  task_total: number
  task_done: number
  attachment_count: number
}

export interface CreateIncidentPayload {
  title: string
  severity: Severity
  template_id?: string
  assigned_to?: string
}

export interface UpdateIncidentPayload {
  title?: string
  severity?: Severity
  status?: IncidentStatus
  executive_summary?: string | null
  notes?: string | null
  lessons_learned?: string | null
  actions_todo?: string | null
  attack_vector?: string[]
  affected_users?: number
  assigned_to?: string | null
}

export const SEVERITY_LABELS: Record<Severity, string> = {
  sev1: 'SEV-1',
  sev2: 'SEV-2',
  sev3: 'SEV-3',
  sev4: 'SEV-4',
}

export const SEVERITY_COLORS: Record<Severity, string> = {
  sev1: 'chip-red',
  sev2: 'chip-yellow',
  sev3: 'chip-blue',
  sev4: 'chip-muted',
}

export const STATUS_LABELS: Record<IncidentStatus, string> = {
  open: 'OPEN',
  contained: 'CONTAINED',
  closed: 'CLOSED',
  monitoring: 'MONITORING',
}

export const STATUS_COLORS: Record<IncidentStatus, string> = {
  open: 'chip-red',
  contained: 'chip-yellow',
  closed: 'chip-green',
  monitoring: 'chip-blue',
}
